// antv/f2/lib/plugin/scroll-bar.js
var Helper = require('../interaction/helper');
var Util = require('../util/common');
var DEFAULT_CFG = {
  mode: 'x',
  xStyle: {
    backgroundColor: 'rgba(202, 215, 239, 0.5)',
    fillerColor: 'rgb(202, 215, 239)',
    size: 4,
    lineCap: 'round',
    offsetX: 0,
    offsetY: 0
  },
  yStyle: {
    backgroundColor: 'rgba(202, 215, 239, 0.5)',
    fillerColor: 'rgb(202, 215, 239)',
    size: 4,
    lineCap: 'round',
    offsetX: 0,
    offsetY: 0
  }
};

module.exports = {
  init: function init(chart) {
    chart.set('_limitRange', {});
    chart.scrollBar = function (cfg) {
      if (cfg === true) {
        cfg = DEFAULT_CFG;
      } else if (Util.isObject(cfg)) {
        cfg = Util.deepMix({}, DEFAULT_CFG, cfg);
      }
      this.set('_scrollBarCfg', cfg);
    };
  },
  clear: function clear(chart) {
    chart.set('_limitRange', {});
  },
  changeData: function changeData(chart) {
    chart.set('_limitRange', {});
  },
  clearInner: function clearInner(chart) {
    var hBar = chart.get('_horizontalBar');
    var vBar = chart.get('_verticalBar');
    hBar && hBar.remove(true);
    vBar && vBar.remove(true);
    chart.set('_horizontalBar', null);
    chart.set('_verticalBar', null);
  },
  afterGeomDraw: function afterGeomDraw(chart) {
    var scrollBarCfg = chart.get('_scrollBarCfg');

    if (!scrollBarCfg) return;
    var data = chart.get('data');
    var plotRange = chart.get('plotRange');
    var backPlot = chart.get('backPlot');
    var canvas = chart.get('canvas');
    var canvasHeight = canvas.get('height');
    var limitRange = chart.get('_limitRange');

    var mode = scrollBarCfg.mode;

    if (Helper.directionEnabled(mode, 'x')) {
      var xStyle = scrollBarCfg.xStyle;
      var xScale = chart.getXScale();
      var xLimitRange = limitRange[xScale.field];
      if (!xLimitRange) {
        xLimitRange = Helper._getLimitRange(data, xScale);
        limitRange[xScale.field] = xLimitRange;
      }

      var currentRange = Helper._getFieldRange(xScale, xLimitRange, xScale.type);
      var horizontalBar = chart.get('_horizontalBar');
      if (horizontalBar) {
        var progressLine = horizontalBar.get('children')[1];

        progressLine.attr({
          x1: Math.max(plotRange.bl.x + plotRange.width * currentRange[0], plotRange.bl.x),
          x2: Math.min(plotRange.bl.x + plotRange.width * currentRange[1], plotRange.br.x)
        });
      } else {
        horizontalBar = backPlot.addGroup({
          className: 'horizontalBar'
        });
        horizontalBar.addShape('line', {
          attrs: {
            x1: plotRange.bl.x,
            y1: canvasHeight - xStyle.size / 2,
            x2: plotRange.br.x,
            y2: canvasHeight - xStyle.size / 2,
            lineWidth: xStyle.size,
            stroke: xStyle.backgroundColor,
            lineCap: xStyle.lineCap
          }
        });
        horizontalBar.addShape('line', {
          attrs: {
            x1: Math.max(plotRange.bl.x + plotRange.width * currentRange[0], plotRange.bl.x),
            y1: canvasHeight - xStyle.size / 2,
            x2: Math.min(plotRange.bl.x + plotRange.width * currentRange[1], plotRange.br.x),
            y2: canvasHeight - xStyle.size / 2,
            lineWidth: xStyle.size,
            stroke: xStyle.fillerColor,
            lineCap: xStyle.lineCap
          }
        });
        chart.set('_horizontalBar', horizontalBar);
      }
    }

    if (Helper.directionEnabled(mode, 'y')) {
      var yStyle = scrollBarCfg.yStyle;
      var yScale = chart.getYScales()[0];

      var yLimitRange = limitRange[yScale.field];
      if (!yLimitRange) {
        yLimitRange = Helper._getLimitRange(data, yScale);
        limitRange[yScale.field] = yLimitRange;
      }

      var _currentRange = Helper._getFieldRange(yScale, yLimitRange, yScale.type);
      var verticalBar = chart.get('_verticalBar');
      if (verticalBar) {
        var _progressLine = verticalBar.get('children')[1];

        _progressLine.attr({
          y1: Math.max(plotRange.tl.y + plotRange.height * _currentRange[0], plotRange.tl.y),
          y2: Math.min(plotRange.tl.y + plotRange.height * _currentRange[1], plotRange.bl.y)
        });
      } else {
        verticalBar = backPlot.addGroup({
          className: 'verticalBar'
        });
        verticalBar.addShape('line', {
          attrs: {
            x1: yStyle.size / 2,
            y1: plotRange.tl.y,
            x2: yStyle.size / 2,
            y2: plotRange.bl.y,
            lineWidth: yStyle.size,
            stroke: yStyle.backgroundColor,
            lineCap: yStyle.lineCap
          }
        });
        verticalBar.addShape('line', {
          attrs: {
            x1: yStyle.size / 2,
            y1: Math.max(plotRange.tl.y + plotRange.height * _currentRange[0], plotRange.tl.y),
            x2: yStyle.size / 2,
            y2: Math.min(plotRange.tl.y + plotRange.height * _currentRange[1], plotRange.bl.y),
            lineWidth: yStyle.size,
            stroke: yStyle.fillerColor,
            lineCap: yStyle.lineCap
          }
        });
        chart.set('_verticalBar', verticalBar);
      }
    }
  }
};
